pkg/codegen/convert/mapper.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+type Mapper interface {
+	GetMapping(provider string) ([]byte, error)
+}
+
+type pluginMapper struct {
+	entries map[string][]byte
+}
+
+func NewPluginMapper(host plugin.Host, key string, mappings []string) (Mapper, error) {
```

Comment:
Is there a reason we prefer `mappings` being a list of file names instead of a `map[string][]byte`. I'm concerned that testing this logic will require writing a bunch of files, which is much more involved than writing a `map`.

Nit: A comment explaining that `key` is the conversion source format would be great. That isn't obvious without reading the function.

Reply:
It's kinda the point of this type (pluginMapper) to do the file and plugin enumeration. If you just want to do mapping with a `map[string][]byte` you can write a new Mapper for that really easy, and that's what I've done in tfconvert based off this code.